File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/tools/drawio/DrawioWriter.kt
/** Writes updated node positions back to draw.io XML document. */
package com.codeintel.tools.drawio

import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.NodeList
import java.io.File
import java.io.StringWriter
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

/** Writes layout results back to the XML document and saves to file. */
object DrawioWriter {

    fun write(doc: Document, graph: DiagramGraph, file: File) {
        val cells = doc.getElementsByTagName("mxCell")
        val allNodes = graph.nodes + graph.containers
        applyPositions(cells, allNodes)
        applyEdgeAnchors(cells, graph)
        applyEdgeRouting(doc, cells, graph)
        saveDocument(doc, file)
    }

    private fun applyPositions(cells: NodeList, nodes: List<DiagramNode>) {
        val nodeMap = nodes.associateBy { it.id }
        for (i in 0 until cells.length) {
            val cell = cells.item(i) as Element
            val id = cell.getAttribute("id")
            val node = nodeMap[id] ?: continue
            val geoms = cell.getElementsByTagName("mxGeometry")
            if (geoms.length == 0) continue
            val geom = geoms.item(0) as Element
            if (geom.getAttribute("as") != "geometry") continue
            geom.setAttribute("x", node.x.format())
            geom.setAttribute("y", node.y.format())
            geom.setAttribute("width", node.width.format())
            geom.setAttribute("height", node.height.format())
        }
    }

    private fun applyEdgeAnchors(cells: NodeList, graph: DiagramGraph) {
        val nodeMap = (graph.nodes + graph.containers).associateBy { it.id }
        for (i in 0 until cells.length) {
            val cell = cells.item(i) as Element
            if (cell.getAttribute("edge") != "1") continue
            val srcId = cell.getAttribute("source") ?: continue
            val tgtId = cell.getAttribute("target") ?: continue
            val src = nodeMap[srcId] ?: continue
            val tgt = nodeMap[tgtId] ?: continue
            val (exitX, exitY) = computeExitPoint(src, tgt)
            val (entryX, entryY) = computeEntryPoint(src, tgt)
            updateEdgeStyle(cell, exitX, exitY, entryX, entryY)
        }
    }

    private fun computeExitPoint(src: DiagramNode, tgt: DiagramNode): Pair<Double, Double> {
        val srcCx = src.x + src.width / 2
        val srcCy = src.y + src.height / 2
        val tgtCx = tgt.x + tgt.width / 2
        val tgtCy = tgt.y + tgt.height / 2
        val dx = tgtCx - srcCx
        val dy = tgtCy - srcCy
        return pickSide(dx, dy)
    }

    private fun computeEntryPoint(src: DiagramNode, tgt: DiagramNode): Pair<Double, Double> {
        val srcCx = src.x + src.width / 2
        val srcCy = src.y + src.height / 2
        val tgtCx = tgt.x + tgt.width / 2
        val tgtCy = tgt.y + tgt.height / 2
        val dx = srcCx - tgtCx
        val dy = srcCy - tgtCy
        return pickSide(dx, dy)
    }

    private fun pickSide(dx: Double, dy: Double): Pair<Double, Double> {
        return if (Math.abs(dx) > Math.abs(dy)) {
            if (dx > 0) Pair(1.0, 0.5) else Pair(0.0, 0.5) // right or left
        } else {
            if (dy > 0) Pair(0.5, 1.0) else Pair(0.5, 0.0) // bottom or top
        }
    }

    private fun updateEdgeStyle(cell: Element, exitX: Double, exitY: Double, entryX: Double, entryY: Double) {
        var style = cell.getAttribute("style") ?: ""
        style = replaceOrAppend(style, "exitX", exitX.format())
        style = replaceOrAppend(style, "exitY", exitY.format())
        style = replaceOrAppend(style, "entryX", entryX.format())
        style = replaceOrAppend(style, "entryY", entryY.format())
        cell.setAttribute("style", style)
    }

    private fun replaceOrAppend(style: String, key: String, value: String): String {
        val regex = Regex("$key=[^;]*")
        return if (regex.containsMatchIn(style)) {
            regex.replace(style, "$key=$value")
        } else {
            if (style.endsWith(";")) "${style}$key=$value;" else "$style;$key=$value;"
        }
    }

    /** Apply orthogonal edge routing — add waypoints where edges cross shapes. */
    private fun applyEdgeRouting(doc: Document, cells: NodeList, graph: DiagramGraph) {
        val routes = EdgeRouter.route(graph)
        if (routes.isEmpty()) return
        for (i in 0 until cells.length) {
            val cell = cells.item(i) as Element
            if (cell.getAttribute("edge") != "1") continue
            val edgeId = cell.getAttribute("id")
            val waypoints = routes[edgeId] ?: continue
            // Set orthogonal edge style
            var style = cell.getAttribute("style") ?: ""
            style = replaceOrAppend(style, "edgeStyle", "orthogonalEdgeStyle")
            style = replaceOrAppend(style, "rounded", "1")
            cell.setAttribute("style", style)
            // Add waypoints as mxPoint array inside mxGeometry
            addWaypoints(doc, cell, waypoints)
        }
    }

    private fun addWaypoints(doc: Document, cell: Element, waypoints: List<Waypoint>) {
        val geoms = cell.getElementsByTagName("mxGeometry")
        val geom = if (geoms.length > 0) geoms.item(0) as Element
            else doc.createElement("mxGeometry").also { it.setAttribute("as", "geometry"); it.setAttribute("relative", "1"); cell.appendChild(it) }
        // Remove existing Array points
        val existing = geom.getElementsByTagName("Array")
        for (i in existing.length - 1 downTo 0) {
            val arr = existing.item(i) as Element
            if (arr.getAttribute("as") == "points") geom.removeChild(arr)
        }
        // Create new Array with mxPoints
        val array = doc.createElement("Array")
        array.setAttribute("as", "points")
        for (wp in waypoints) {
            val point = doc.createElement("mxPoint")
            point.setAttribute("x", wp.x.format())
            point.setAttribute("y", wp.y.format())
            array.appendChild(point)
        }
        geom.appendChild(array)
    }

    private fun saveDocument(doc: Document, file: File) {
        val transformer = TransformerFactory.newInstance().newTransformer()
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
        val writer = StringWriter()
        transformer.transform(DOMSource(doc), StreamResult(writer))
        file.writeText(writer.toString())
    }

    private fun Double.format(): String {
        val rounded = Math.round(this).toDouble()
        return if (rounded == this) rounded.toLong().toString() else "%.1f".format(this)
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/drawio/drawio_writer.py
"""Writes updated node positions back to draw.io XML document.

Approach: reposition nodes to grid, REMOVE all manual waypoints and
exit/entry anchors — let draw.io's built-in edge router handle connectors.
"""

import re
import xml.etree.ElementTree as ET
from .drawio_parser import DiagramGraph, DiagramNode


def write_layout(tree: ET.ElementTree, graph: DiagramGraph, file_path: str) -> None:
    """Write layout results back to XML and save to file."""
    root = tree.getroot()
    all_nodes = graph.nodes + graph.containers
    cells = root.findall(".//" + "mxCell")
    _apply_positions(cells, all_nodes)
    _clean_edge_routing(cells)
    tree.write(file_path, encoding="unicode", xml_declaration=True)


def _apply_positions(cells: list, nodes: list) -> None:
    """Update x, y, width, height in mxGeometry elements."""
    node_map = {n.id: n for n in nodes}
    for cell in cells:
        cid = cell.get("id", "")
        node = node_map.get(cid)
        if not node:
            continue
        geom = cell.find("mxGeometry")
        if geom is None or geom.get("as") != "geometry":
            continue
        geom.set("x", _fmt(node.x))
        geom.set("y", _fmt(node.y))
        geom.set("width", _fmt(node.width))
        geom.set("height", _fmt(node.height))


def _clean_edge_routing(cells: list) -> None:
    """Remove manual waypoints and anchor overrides from edges.

    Draw.io's built-in edge router handles routing automatically.
    Manual waypoints and exit/entry anchors often cause ugly results.
    """
    for cell in cells:
        if cell.get("edge") != "1":
            continue
        # Remove manual exit/entry anchor points from style
        style = cell.get("style", "")
        style = re.sub(r"exitX=[^;]*;?", "", style)
        style = re.sub(r"exitY=[^;]*;?", "", style)
        style = re.sub(r"exitDx=[^;]*;?", "", style)
        style = re.sub(r"exitDy=[^;]*;?", "", style)
        style = re.sub(r"entryX=[^;]*;?", "", style)
        style = re.sub(r"entryY=[^;]*;?", "", style)
        style = re.sub(r"entryDx=[^;]*;?", "", style)
        style = re.sub(r"entryDy=[^;]*;?", "", style)
        # Clean trailing semicolons
        style = re.sub(r";+", ";", style).strip(";")
        if style:
            style += ";"
        cell.set("style", style)
        # Remove manual waypoints (Array as="points")
        geom = cell.find("mxGeometry")
        if geom is not None:
            for arr in list(geom.findall("Array")):
                if arr.get("as") == "points":
                    geom.remove(arr)


def _fmt(val: float) -> str:
    """Format number: integer if whole, else 1 decimal."""
    rounded = round(val)
    if rounded == val:
        return str(int(rounded))
    return f"{val:.1f}"
